teps/0060-remote-resource-resolution.md

```diff
@@ -0,0 +1,346 @@
+
+## Summary
+
+This TEP describes problems around fetching and running Tasks and Pipelines
+("Tekton resources") from remote locations like image registries, git
+repositories, internal catalogs, cloud storage buckets, other clusters or
+namespaces, and so on. As part of the problem statement this TEP also describes
+some of the constraints, goals to aim for during design, as well as some
+possible extensions that could stem from a solution to these problems.
+
+## Key Terms
+
+This section defines some key terms used throughout the rest of this doc.
+
+- "Remote": Any location that stores Tekton resources outside of the cluster
+  (or namespace in multi-tenant setups) that the user's pipelines are running
+  in. These could be: OCI registries, git repos and other version control
+  systems, other namespaces, other clusters, cloud buckets, an organization's
+  proprietary internal catalog, or any other storage system.
+- "Remote Resource": The YAML file or other representation of a Tekton resource
+  that lives in the Remote.
+- "Resolver": A program or piece of code that knows how to interpret a
+  reference to a remote resource and fetch it for execution as part of the
+  user's TaskRun or Pipeline.
+- "Resource Resolution": The act of taking a reference to a Tekton resource in
+  a Remote, fetching it, and returning its content to the Tekton Pipelines
+  reconcilers.
+
+## Motivation
+
+Pipelines currently provides support for resources to be run from two different
+locations: a Task or Pipeline can be run from the same cluster or (when the
+[`enable-tekton-oci-bundles`
+flag](https://github.com/tektoncd/pipeline/blob/main/config/config-feature-flags.yaml)
+is `"true"`) from an image registry hosting [Tekton
+Bundles](https://tekton.dev/docs/pipelines/pipelines/#tekton-bundles).
+
+This existing support has a few problems:
+
+1. For resources in a cluster the level of control afforded to operators is
+   extremely narrow: Either a resource exists in the cluster or it does not. To
+   add new Tasks and Pipelines operators have to manually (or via automation)
+   sync them in (`kubectl apply -f git-clone.yaml`).
+2. For resources in a Tekton Bundle registry, operators don't have a choice
+   over what kind of storage best suits their teams: It's either a registry or
+   it's back to manually installing Tasks and Pipelines. This might not jive
+   with an org's chosen storage. Put another way: [why can't we just keep our
+   Tasks in git?](https://github.com/tektoncd/pipeline/issues/2298)
+3. Pipeline authors have to document the Tasks their Pipeline depends on
+   out-of-band. A [common example of
+   this](https://github.com/tektoncd/pipeline/discussions/3822) is a build
+   Pipeline that needs `git-clone` to be installed from the open source catalog
+   before the Pipeline can be run.
+   [TEP-0053](https://github.com/tektoncd/community/pull/352) is also exploring
+   ways of encoding this information as part of pipelines in the catalog.
```

Comment:
i wonder if it would make sense to add a requirement around catalog interoperability - i.e. if we go this route, and folks want to add Pipelines to the catalog, they're going to need a reliable way to reference Tasks - maybe we need to have some minimum number of resource resolution mechanisms that work out of the box (tho folks could disable them if needed)? im thinking oci bundles for sure - maybe git as well...

Reply:
Linking relevant discussion on [TEP-0053](https://github.com/tektoncd/community/pull/352#discussion_r604136070) as well.

Reply:
One of the Goals already describes this a little bit I think:

```
- Establish a common syntax that tool and platform creators can use to record,
  as part of a pipeline or run, the remote location that Tekton resources
  should be fetched from.
```

To record the requirement of providing bundle and in-cluster support by default I've added a new req:

```
- At minimum we should provide resolver implementations for Tekton Bundles and
  in-cluster resources. These can be provided out-of-the-box just as they are
  today. The code for these resolvers can either be hosted as part of
  Pipelines, in the Catalog, or in a new repo under the `tektoncd` GitHub org.
```

And then to capture adding git support I've added this one too:

```
- Add new support for resolving resources from git via this mechanism. This
  could be provided out of the box too but we don't _have to_ since this
  doesn't overlap with concerns around backwards compatibility in the same way
  that Tekton Bundles and in-cluster support might.
```